File: benchmark/synduce/constraints/sortedlist/count_lt.f
Inductive List = elt Int | cons {Int, List};

is_sorted =
  let aux = fix (
    \f: Int -> List -> Bool. \pre: Int. \xs: List.
    match xs with
      elt x -> <= pre x
    | cons {h, t} -> and (<= pre h) (f h t)
    end
  ) in \xs: List.
  match xs with
    elt x -> true
  | cons {h, t} -> aux h t
  end;

@Input winp: Int;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt a -> if < a winp then 1 else 0
  | cons {h, t} ->
    + (if < h winp then 1 else 0) (f t)
  end
);

target = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    elt w -> xs
  | cons {h, t} ->
    if < h winp then cons {h, f t}
    else xs /*Avoid recursions*/
  end
);

main = \xs: List. if is_sorted xs then spec (target xs) else 0;

File: ml-benchmark/synduce/constraints/sortedlist/count_lt.ml
@Input val winp: int
type list = Elt of int | Cons of int * list

let rec aux pre xs =
  match xs with
  | Elt x -> pre <= x
  | Cons (h, t) -> (pre <= h) && aux h t

let is_sorted xs =
  match xs with
  | Elt x -> true
  | Cons (h, t) -> aux h t

let rec spec xs =
  match xs with
  | Elt a -> if a < winp then 1 else 0
  | Cons (h, t) ->
      let c = if h < winp then 1 else 0 in
      c + spec t

val target: list -> list compress
let rec target xs =
  match xs with
  | Elt w -> xs
  | Cons (h, t) ->
      if h < winp then Cons (h, target t) else xs

let program xs =
  if is_sorted xs then spec (target xs) else 0